File: db-tester-junit-spring-boot-starter/src/main/java/io/github/seijikohara/dbtester/junit/spring/boot/autoconfigure/SpringBootDatabaseTestExtension.java
package io.github.seijikohara.dbtester.junit.spring.boot.autoconfigure;

import io.github.seijikohara.dbtester.api.config.Configuration;
import io.github.seijikohara.dbtester.junit.jupiter.extension.DatabaseTestExtension;
import java.util.Optional;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.ExtensionContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.ApplicationContext;
import org.springframework.test.context.junit.jupiter.SpringExtension;

/**
 * Spring Boot-aware database test extension that automatically registers DataSources.
 *
 * <p>This extension extends {@link DatabaseTestExtension} and adds automatic DataSource
 * registration from the Spring {@link ApplicationContext}. When the {@code
 * db-tester.auto-register-data-sources} property is set to {@code true} (the default), all
 * Spring-managed {@link javax.sql.DataSource} beans are automatically registered with the {@link
 * io.github.seijikohara.dbtester.api.config.DataSourceRegistry} before any test execution.
 *
 * <p>For custom DataSource registration (e.g., multiple DataSources with specific names), use the
 * parent class {@link DatabaseTestExtension} directly and perform manual registration in
 * {@code @BeforeAll}.
 *
 * @see DatabaseTestExtension
 * @see DataSourceRegistrar
 * @see DbTesterJUnitAutoConfiguration
 */
public class SpringBootDatabaseTestExtension extends DatabaseTestExtension
    implements BeforeAllCallback {

  /** Logger for tracking automatic DataSource registration. */
  private static final Logger logger =
      LoggerFactory.getLogger(SpringBootDatabaseTestExtension.class);

  /**
   * Creates a new Spring Boot database test extension.
   *
   * <p>This constructor is called by JUnit Jupiter when the extension is registered via
   * {@code @ExtendWith}.
   */
  public SpringBootDatabaseTestExtension() {
    super();
  }

  /**
   * Automatically registers Spring-managed DataSources and Configuration before all tests.
   *
   * <p>This method:
   *
   * <ol>
   *   <li>Retrieves the Spring {@link ApplicationContext} using {@link SpringExtension}
   *   <li>Registers the Spring-managed {@link Configuration} bean with the extension context
   *   <li>Checks if {@link DataSourceRegistrar} is available in the context
   *   <li>If available and auto-registration is enabled, registers all DataSources with the {@link
   *       io.github.seijikohara.dbtester.api.config.DataSourceRegistry}
   * </ol>
   *
   * <p>If the Spring context is not available (e.g., non-Spring tests) or if {@link
   * DataSourceRegistrar} is not configured, this method silently skips automatic registration,
   * allowing the extension to work in both Spring and non-Spring environments.
   *
   * @param context the extension context for the test class
   */
  @Override
  public void beforeAll(final ExtensionContext context) {
    try {
      final var applicationContext = SpringExtension.getApplicationContext(context);
      registerConfigurationFromContext(context, applicationContext);
      registerDataSourcesFromContext(context, applicationContext);
    } catch (final IllegalStateException e) {
      // Spring context not available - this is expected for non-Spring tests
      logger.debug(
          "Spring ApplicationContext not available, skipping automatic DataSource registration: {}",
          e.getMessage());
    }
  }

  /**
   * Registers Configuration from the Spring ApplicationContext.
   *
   * <p>If a Configuration bean is available in the Spring context, it will be used by the test
   * extension. This allows configuration to be customized via application.properties.
   *
   * @param context the extension context
   * @param applicationContext the Spring application context
   */
  private void registerConfigurationFromContext(
      final ExtensionContext context, final ApplicationContext applicationContext) {

    Optional.of(applicationContext)
        .filter(appContext -> appContext.containsBean("dbTesterConfiguration"))
        .map(appContext -> appContext.getBean("dbTesterConfiguration", Configuration.class))
        .ifPresentOrElse(
            configuration -> {
              setConfiguration(context, configuration);
              logger.debug(
                  "Registered Spring-managed Configuration with database testing framework");
            },
            () ->
                logger.debug(
                    "Configuration bean not found in ApplicationContext, using default configuration"));
  }

  /**
   * Registers DataSources from the Spring ApplicationContext.
   *
   * @param context the extension context
   * @param applicationContext the Spring application context
   */
  private void registerDataSourcesFromContext(
      final ExtensionContext context, final ApplicationContext applicationContext) {

    Optional.of(applicationContext)
        .filter(appContext -> appContext.containsBean("dataSourceRegistrar"))
        .ifPresentOrElse(
            appContext -> registerDataSources(context, appContext),
            () ->
                logger.debug(
                    "DataSourceRegistrar bean not found in ApplicationContext, skipping automatic DataSource registration"));
  }

  /**
   * Registers DataSources using the registrar bean.
   *
   * @param context the extension context
   * @param applicationContext the Spring application context
   */
  private void registerDataSources(
      final ExtensionContext context, final ApplicationContext applicationContext) {

    final var registrar = applicationContext.getBean(DataSourceRegistrar.class);
    final var registry = getRegistry(context);

    logger.info("Automatically registering Spring DataSources with database testing framework");
    registrar.registerAll(registry);
    logger.info("Automatic DataSource registration completed");
  }
}

File: db-tester-kotest-spring-boot-starter/src/main/kotlin/io/github/seijikohara/dbtester/kotest/spring/boot/autoconfigure/SpringBootDatabaseTestExtension.kt
package io.github.seijikohara.dbtester.kotest.spring.boot.autoconfigure

import io.github.seijikohara.dbtester.api.config.Configuration
import io.github.seijikohara.dbtester.api.config.DataSourceRegistry
import io.github.seijikohara.dbtester.kotest.extension.DatabaseTestExtension
import io.kotest.core.extensions.SpecExtension
import io.kotest.core.extensions.TestCaseExtension
import io.kotest.core.spec.Spec
import io.kotest.core.test.TestCase
import io.kotest.engine.test.TestResult
import org.slf4j.LoggerFactory
import org.springframework.context.ApplicationContext
import org.springframework.test.context.TestContextManager

/**
 * Spring Boot-aware database test extension for Kotest that automatically registers DataSources.
 *
 * This extension wraps [DatabaseTestExtension] and adds automatic DataSource registration
 * from the Spring [ApplicationContext]. When the `db-tester.auto-register-data-sources`
 * property is set to `true` (the default), all Spring-managed [javax.sql.DataSource] beans
 * are automatically registered with the [DataSourceRegistry] before any test execution.
 *
 * **Usage:**
 * ```kotlin
 * @SpringBootTest
 * class MyTest : AnnotationSpec() {
 *
 *     init {
 *         extensions(SpringBootDatabaseTestExtension())
 *     }
 *
 *     @Test
 *     @Preparation
 *     @Expectation
 *     fun `should verify database state`() {
 *         // test implementation
 *     }
 * }
 * ```
 *
 * @see DatabaseTestExtension
 * @see DataSourceRegistrar
 * @see DbTesterKotestAutoConfiguration
 */
class SpringBootDatabaseTestExtension :
    SpecExtension,
    TestCaseExtension {
    /** Companion object containing class-level logger. */
    companion object {
        private val logger = LoggerFactory.getLogger(SpringBootDatabaseTestExtension::class.java)
    }

    private var registry: DataSourceRegistry? = null
    private var configuration: Configuration? = null
    private var delegate: DatabaseTestExtension? = null
    private var applicationContext: ApplicationContext? = null

    /**
     * Called before the spec is instantiated.
     *
     * Automatically registers Spring-managed DataSources and Configuration.
     *
     * @param spec the spec instance
     */
    override suspend fun intercept(
        spec: Spec,
        execute: suspend (Spec) -> Unit,
    ): Unit = initializeFromSpec(spec).let { execute(spec) }

    /**
     * Intercepts test case execution to handle preparation and expectation phases.
     *
     * @param testCase the test case being executed
     * @param execute the function to execute the test case
     * @return the test result
     */
    override suspend fun intercept(
        testCase: TestCase,
        execute: suspend (TestCase) -> TestResult,
    ): TestResult =
        (delegate ?: initializeFromSpec(testCase.spec).let { delegate!! })
            .intercept(testCase, execute)

    /**
     * Initializes the extension from the Spec instance.
     *
     * @param spec the spec instance
     */
    private fun initializeFromSpec(spec: Spec): Unit =
        when (delegate) {
            null ->
                getApplicationContext(spec).also { applicationContext = it }.let { context ->
                    registerConfigurationFromContext(context)
                    registerDataSourcesFromContext(context)
                    delegate =
                        DatabaseTestExtension(
                            registryProvider = { registry!! },
                            configurationProvider = { configuration ?: Configuration.defaults() },
                        )
                }
            else -> Unit
        }

    /**
     * Retrieves the ApplicationContext from the Spec using Spring's TestContextManager.
     *
     * @param spec the spec instance
     * @return the application context
     */
    private fun getApplicationContext(spec: Spec): ApplicationContext =
        TestContextManager(spec::class.java)
            .also { it.prepareTestInstance(spec) }
            .testContext
            .applicationContext

    /**
     * Registers Configuration from the Spring ApplicationContext.
     *
     * @param applicationContext the application context
     */
    private fun registerConfigurationFromContext(applicationContext: ApplicationContext): Unit =
        when {
            applicationContext.containsBean("dbTesterConfiguration") -> {
                configuration = applicationContext.getBean("dbTesterConfiguration", Configuration::class.java)
                logger.debug("Registered Spring-managed Configuration with database testing framework")
            }
            else -> logger.debug("Configuration bean not found in ApplicationContext, using default configuration")
        }

    /**
     * Registers DataSources from the Spring ApplicationContext.
     *
     * @param applicationContext the application context
     */
    private fun registerDataSourcesFromContext(applicationContext: ApplicationContext): Unit =
        when {
            applicationContext.containsBean("dbTesterDataSourceRegistry") -> {
                registry = applicationContext.getBean("dbTesterDataSourceRegistry", DataSourceRegistry::class.java)
                logger.debug("Using Spring-managed DataSourceRegistry")
            }
            applicationContext.containsBean("dataSourceRegistrar") -> {
                registry = DataSourceRegistry()
                logger.info("Automatically registering Spring DataSources with database testing framework")
                applicationContext.getBean(DataSourceRegistrar::class.java).registerAll(registry!!)
                logger.info("Automatic DataSource registration completed")
            }
            else -> {
                registry = DataSourceRegistry()
                logger.debug("DataSourceRegistrar bean not found, using empty registry")
            }
        }
}
